clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/PatchExecutor.scala

```diff
@@ -0,0 +1,146 @@
+import org.broadinstitute.clio.JsonUtils.JsonOps
+
+class PatchExecutor[CI <: ClioIndex](patchCommand: PatchCommand[CI]) extends Executor {
+
+  import patchCommand.index.implicits._
+  import Executor.SourceMonadOps
+
+  private type DocumentKey = patchCommand.index.KeyType
+
+  private val queryAllFile: File =
+    File.newTemporaryFile().deleteOnExit().write(Json.obj().pretty(implicitly))
+
+  /**
+    * Build a stream which, when pulled, will communicate with the clio-server
+    * to update its records about some metadata, potentially performing IO
+    * operations in the process.
+    */
+  override def execute(
+    webClient: ClioWebClient,
+    ioUtil: IoUtil
+  ): Source[Json, NotUsed] = {
+    // Read the metadata then turn into Json so that we fail fast.
+    ioUtil
+      .readMetadata(patchCommand.index)(patchCommand.metadataLocation)
+      .map(_.asJson.dropNulls)
+      .flatMap(Source.repeat)
+      // zip new metadatas with documents
+      .zip(
+        webClient
+          .jsonFileQuery(patchCommand.index)(queryAllFile)
+          .map(_.dropNulls)
+      )
+      // Munge metadata for documents
+      .map {
+        case (newMetadataJson, documentJson) =>
+          mungeMetadata(newMetadataJson, documentJson)
+      }
+      // filter out things that already have no new metadata to upsert
+      .filterNot {
+        case (mergedJson, oldMetadataJson, _) =>
+          mergedJson.equals(oldMetadataJson)
+      }
+      // Upsert the stuff!
+      .flatMapConcat {
+        case (mergedMetadataJson, oldMetadataJson, docKey) =>
+          upsertPatch(mergedMetadataJson, oldMetadataJson, docKey, webClient)
+      }
+  }
+
+  /**
+    * Munge the patch metadata with the document metadata to get a new metadata json
+    * @param patchMetadataJson The new metadata to insert into the index
+    * @param documentJson The old document to insert the metadata into
+    * @return A tuple containing the munged metadata, the old metadata, and the key for this document
+    */
+  private def mungeMetadata(
+    patchMetadataJson: Json,
+    documentJson: Json
+  ): (Json, Json, DocumentKey) = {
+    val oldMetadataJson = getMetadataJsonFromDocument(documentJson)
+    val indexKey = getKeyFromDocument(documentJson)
+    (
+      patchMetadataJson.deepMerge(oldMetadataJson),
+      oldMetadataJson,
+      indexKey
+    )
+  }
+
+  /**
+    * Upsert the new metadata into the document.
+    * This method strips any existing keys out of the metadata to be upserted in order to reduce traffic.
+    * @param mergedMetadataJson The merged metadata containing patch and existing metadata
+    * @param oldMetadataJson The old, preexisting metadata
+    * @param docKey The key for this metadata
+    * @param webClient A ClioWebClient to upsert with.
+    * @return
+    */
+  private def upsertPatch(
+    mergedMetadataJson: Json,
+    oldMetadataJson: Json,
+    docKey: DocumentKey,
+    webClient: ClioWebClient
```

Comment:
It seems odd to me that `ClioWebClient` and `IoUtil` aren't members of `Executor`, e.g. provided at construction time, since they don't vary over the lifetime of the executor, and are needed for the executor to work.